### PR TITLE
First integration of JOS3 model and pythermalcomfort

### PR DESCRIPTION
**Pull request by AkihisaNomoto**
Hello! 

This is the first integration of JOS3 model and pythermalcomfort. The JOS-3 model code is basically copied from here (https://github.com/TanabeLab/JOS-3) but I newly added some description about the model and comments to the code for readers to easily understand.

Main changes are as follows;
- Added the main code of JOS3 model (class JOS3) into models.py. Please make sure the code works properly on your computer.
- **The nomenclature is the same as the original code for now to keep consistency with the paper (https://www.sciencedirect.com/science/article/pii/S0378778820333612), but if you feel uncomfortable with that we can change it to keep consistency with the other nomenclature in pythermalcomfort.** 
- **Probably you automatically sync the model's docstrings and pythermalcomfort documentaion but I am not sure JOS3 model's docstrings can be properly synced. Please comfirm this and modify as needed.**
- Generate 'jos3' folder under 'pythermalcomfort' package. This folder has some python codes only for JOS3 model. 
- Added an example code for JOS3 model called 'cals_jos3.py' in 'example' folder. Example outputs are in a folder named 'jos3_output_example'.
- Changed _body_surface_area_ function in 'utilities.py'. Added some additional formula for calculating body surface area, not only dubois's formula.
- Added _add_prompt_to_code_ function in 'utilities.py' to make an example code in docstrings.

This is just an initial commit and I will modify the code for JOS3 model as needed.

Thanks,
Aki

### CODE REVIEW

**Review by FedericoTartarini — COMMENTED** [review]
Thank you so much for the great work. I am extremely happy that we are integrating the jos3 model inside pythermalcomfort. I am sure the scientific community will greatly benefit from that. The code you wrote is perfect. I will check the formatting before publishing and the docstrigs but overall I do not see any major issues.

Few comments below:
* please remove the pmv functions since it is a duplicate we can use the one in pythermalcomfort
* we will need to write some tests to make sure we get consistent results. I can do that, we just need to define a few simulations we want to test.
* can we delete the jos3 folder inside `pythermalcomfort` and merge the `comfmod`, `construction`, `matrix`, and `params` files into one and call it jo3_functions.py This will ensure that all the jos3 supporting functions are in one file.
* I left some comments in the code but they are kind of duplicates of the comments here
* It would be nice to keep the variable names consistent with pythermalcomfort. For example RH should be rh. I can help with that if you agree.

Aside from these minor issues all the rest looks great and I can then merge the pull request.

File: src/pythermalcomfort/jos3/comfmod.py
Comment (AkihisaNomoto) [reply src/pythermalcomfort/jos3/comfmod.py:43]: OK. I just left this function to adoid Circular Import Error between model.py and comfmod.py. I will remove the function to calculate PMV and move the function to calculate the operative temperature when PMV=0 somewhere else.

File: src/pythermalcomfort/models.py
Comment (AkihisaNomoto) [reply src/pythermalcomfort/models.py:3377]: OK. I will add the order.

File: src/pythermalcomfort/jos3/thermoregulation.py
Comment (AkihisaNomoto) [reply src/pythermalcomfort/jos3/thermoregulation.py:14]: Right. I will at least add a comment about which body part is reffered to each element. When we were writing the original code of JOS3, we hesiteted to use a dictionary because I feel Python is not good at iteration calculation (for loop) and it takes time to calculate. However, using numpy may solve this problem (reference: https://numpy.org/devdocs/user/absolute_beginners.html), we haven't checked the calculation speed though. Let's discuss this later.

File: src/pythermalcomfort/models.py
Comment (AkihisaNomoto) [reply src/pythermalcomfort/models.py:3354]: OK. Then let's change the variables' names.

File: src/pythermalcomfort/jos3/comfmod.py
Comment (FedericoTartarini) [reply src/pythermalcomfort/jos3/comfmod.py:43]: great, thank you for implementing this.

File: src/pythermalcomfort/jos3/thermoregulation.py
Comment (FedericoTartarini) [reply src/pythermalcomfort/jos3/thermoregulation.py:14]: I agree with you that we should use Numpy to perform the calculations. What we could do is to allow the user to pass a dictionary. We then convert the dictionary to a Numpy array and perform the calculations. This will not negatively affect the speed of the code while at the same time it will ensure that the user enters the right inputs.

We could also use @dataclasses instead of dictionaries to make sure the code is even more robust. Below a simple example.

```
from dataclasses import dataclass

@dataclass
class BodyParts:
    arm: float
    head: float
```

Comment (AkihisaNomoto) [reply src/pythermalcomfort/jos3/thermoregulation.py]: I have changed the code to allow the users to pass a dictionary as well as interger value, list, numpy array data format.

File: src/pythermalcomfort/models.py
Comment (AkihisaNomoto) [reply src/pythermalcomfort/models.py]: Changed.

File: src/pythermalcomfort/models.py
Comment (AkihisaNomoto) [reply src/pythermalcomfort/models.py]: Added a note at the top of the code saying that numpy arrays in the code correspond to the body parts.

**Comment by AkihisaNomoto** [comment]
Hi,

Thanks for all your commnets.

Changes are as follows as we discussed yesterday;
- Chnaged the variables' names and make them consistent with the other functions in pythermalcomfort.
- Changed the code to allow the users to pass a dictionary as well as interger value, list, numpy array data format.

New chnages since yesterday is;
- Added csv file for local clothing insulations of typical cloting ensembles in pythermalcomfort folder for JOS-3 model input. This database is based on the study by Juyoun et al. (https://escholarship.org/uc/item/18f0r375) and by me (https://doi.org/10.1002/2475-8876.12124). I added a function to convert this csv file to dictionary and added the converted dictionary to `utilities.py`.

Pelase feel free to edit as you like.

Thank you again for adding me to pythermalcomfort co-maintainers. I am happy to internaly work on developing this model.

Akihisa

**Comment by AkihisaNomoto** [comment]
One more thing - I added a function to get the current version of pythermalcomfort from `__init__.py` and print it in the output of JOS-3 model to let the users know what version they are using. I think this is important because JOS-3 model can be further developed.

File: src/pythermalcomfort/local_clo_summary.csv
Comment (FedericoTartarini) [inline src/pythermalcomfort/local_clo_summary.csv:1]: I belive that this file should be moved to another folder and not inside `src/pythermalcomfort`. Is this a file you need to run the JOS model or is an example file?

File: src/pythermalcomfort/models.py
Comment (FedericoTartarini) [inline src/pythermalcomfort/models.py:282]: why did you remove the pmv_ppd function from here?

File: src/pythermalcomfort/models.py
Comment (FedericoTartarini) [inline src/pythermalcomfort/models.py:1670]: could you please remove these formatting changes? I formatted the documentation with a Python package and unless is creating issues I would like to keep it as it is. If you reformatted since you spotted some issues in the docs please let me know and I am happy to change it.

File: src/pythermalcomfort/models.py
Comment (FedericoTartarini) [inline src/pythermalcomfort/models.py:25]: move this import to the top since are python libraries and then we can import all the files and functions from pythermalcomfort

File: src/pythermalcomfort/shared_functions.py
Comment (FedericoTartarini) [inline src/pythermalcomfort/shared_functions.py:10]: since this is a function that is only used by the JOS model maybe we could include it in one of the files in the `jos3_functions` folder

File: src/pythermalcomfort/shared_functions.py
Comment (FedericoTartarini) [inline src/pythermalcomfort/shared_functions.py:49]: same as the previous comment

File: src/pythermalcomfort/utilities.py
Comment (FedericoTartarini) [inline src/pythermalcomfort/utilities.py:546]: these dictionaries are also only specific to the JOS3 model so we may want to move them to the `jos3_funcitons` folder.

**Review by FedericoTartarini — COMMENTED** [review]
Thank you for implementing all these changes. I left a few comments. Sorry that I missed them, please send me an email next time you want me to review the comments since I do not check GitHub regularly.

File: src/pythermalcomfort/local_clo_summary.csv
Comment (AkihisaNomoto) [reply src/pythermalcomfort/local_clo_summary.csv:1]: OK. This file is just a summary table for local clothing insulation for each body part as well as the whole body. This is not directry related to the JOS model or an example file, is just a reference for the users to input local clothing insulation if they wanted to simulate a ditailed condition. I added the same data as a dictinary format in `utilities.py` so that the user can get the values from there, but it is a long dictionary and hard to see. That is why I put a csv file. I newly made a new folder called `others` but you can remove if you do not think this is necessary.

**Comment by AkihisaNomoto** [comment]
Sorry that I messed up a couple of lines of code. I think I could modify it and now all set.

File: others/local_clo_summary.csv
Comment (FedericoTartarini) [inline others/local_clo_summary.csv:1]: Ideally, I would like to keep pythermalcomfort structure clean, this new folder may not be necessary, since these information are duplicated of the values that are in the utilities file.

I would suggest deleting this folder and perhaps move this file to another location. I cannot think of an appropriate location at the moment since this information are duplicated. We could move this file in `src/pythermalcomfort/jo3_functions` what do you think? The other issue is that unless we clearly describe how to use this file in the documentation only you and I will know about its existence and no one else will use it.

File: src/pythermalcomfort/jos3_functions/jos3_utilities.py
Comment (FedericoTartarini) [inline src/pythermalcomfort/jos3_functions/jos3_utilities.py:1]: no need to call this file jos3_utilities since being inside the jos3_functions folder it is clear that it is related to the jos3 mode. please change the name to utilities.

File: src/pythermalcomfort/jos3_functions/params.py
Comment (FedericoTartarini) [inline src/pythermalcomfort/jos3_functions/params.py:1]: could you please change this file name to parameters.py

**Review by FedericoTartarini — COMMENTED** [review]
Great job, I have added two few more comments, once addressed I will merge the pull request. Thank you so much for your help and for working on this.

**Comment by AkihisaNomoto** [comment]
Thanks for your comment. I modified the code based on your feedback.

**Review by FedericoTartarini — APPROVED** [review]

**Pull request merged**
